### PR TITLE
[BRE-1125] Support Hashicorp Vault versioned key / value store

### PR DESCRIPTION
**Pull request by ragurney**
Refactors the secret logic in Samson to support Hashicorp's versioned Key / Value store (KV Secrets Engine - Version 2).
 
#### Notable details about this implementation:
1. This is currently using a [forked version](https://github.com/zendesk/vault-ruby) of [vault-ruby](https://github.com/hashicorp/vault-ruby) containing changes introduced by [this PR](https://github.com/hashicorp/vault-ruby/pull/188). Once that pull request has been merged, the Gemfile should be updated to point to the original Gem again.
2. To support the new API, two new constructs -- `logical(:versioned)` and `logical(:unversioned)` -- were introduced which have different method signatures out of necessity but require the same data. To have one common API that all code could use within Samson to interact with the Vault client, this PR introduces wrappers for both the client and logicals which cleans up the code and reduces a bunch of branching logic. This also gives us the benefit of having more flexibility in the future for adding custom logic / refactoring the existing behavior as needed. 

#### Changes to the UI:
screen shot 2018-12-04 at 11 19 12 am

/cc @zendesk/samson

### Tasks
 - [ ] :+1: from team
 - [ ] Thoroughly test on staging

### References
 - Jira link: [BRE-1125](https://zendesk.atlassian.net/browse/BRE-1125)

### Risks
- Level: High - this should be thoroughly tested in staging to ensure secrets are functioning normally, as this is core Samson logic for deployment.

### CODE REVIEW

**Review by grosser — APPROVED** [review]
looks like you snuck a big refactor into the "add a boolean" PR ... 2 PRs would have been better

File: app/controllers/vault_servers_controller.rb [hunk 1]
```diff
@@ -45,7 +45,7 @@ def destroy
   private
 
   def server_params
-    params.require(:vault_server).permit(:name, :address, :token, :ca_cert, :tls_verify)
+    params.require(:vault_server).permit(:name, :address, :token, :using_versioned_kv, :ca_cert, :tls_verify)
```
Comment (grosser) [inline app/controllers/vault_servers_controller.rb]: just `versioned_kv` would have worked too

File: lib/samson/periodical.rb [hunk 2]
```diff
@@ -15,21 +15,30 @@ module Periodical
 
     class ExceptionReporter
       def initialize(task_name)
+        puts "DEBUG:: NEW EXCEPTION REPORTER #{task_name}" if $debug_messages
         @task_name = task_name
       end
 
       def update(time, _result, exception)
+        puts "DEBUG:: IN EXCEPTION REPORTER UPDATE #{exception.inspect}" if $debug_messages
```
Comment (grosser) [inline lib/samson/periodical.rb]: periodical fix should be it's own PR ?

Comment (ragurney) [reply lib/samson/periodical.rb]: Yeah it should and will be, I'm just using this PR to try to debug why those tests are failing. I can't merge this anyway until I can test it out on staging :/

File: lib/samson/secrets/vault_client_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+require 'vault'
+
+module Samson
+  module Secrets
+    class VaultClientWrapper < Vault::Client
+      # TODO: make these configurable via env vars
```
Comment (grosser) [inline lib/samson/secrets/vault_client_wrapper.rb]:
```suggestion
      # TODO: make these configurable via env vars and send env vars to samson_secret_puller
```

File: lib/samson/secrets/vault_server.rb [hunk 4]
```diff
@@ -1,27 +1,9 @@
...
-    end
-    keys.each { |k| k.slice!(0, path.size) } if root
-    keys
-  end
-end
-
 module Samson
   module Secrets
     class VaultServer < ActiveRecord::Base
-      PREFIX = ENV['VAULT_PREFIX'] || 'secret/apps/'
```
Comment (grosser) [inline lib/samson/secrets/vault_server.rb]: remove all docs we had on this if any

**Review by grosser — APPROVED** [review]
pretty ugly but looks about right :)

a PR to vault-ruby to support prefix and vault would make samson code a lot less messy

**Comment by ragurney** [comment]
Now blocked by the status of [#188](https://github.com/hashicorp/vault-ruby/pull/188) :(

**Comment by ragurney** [comment]
Superseded by #3117
